### PR TITLE
Dockerizing

### PR DESCRIPTION
**Pull request by jared-cain**
Dockerfile, docker-compose.yml, and addtiontal reqs for compute/worker

### CODE REVIEW

File: DOCKER_README.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+## Docker Compose User Guide
```
Comment (ckcollab) [inline DOCKER_README.md]: Awesome thanks for all the detailed instructions! Doing a brief code review now, will dig into this more ASAP

File: DOCKER_README.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+In order to get started, make a copy of _settings/local_sample.py_ and call it *settings/local.py* as described in the [CodaLab Wiki](https://github.com/codalab/codalab/wiki)
```
Comment (ckcollab) [inline DOCKER_README.md]: Maybe add a link to install Docker?

File: DOCKER_README.md (hunk 1 above)
Comment (ckcollab) [inline DOCKER_README.md]: Should we maybe move this to a /docs/ folder or something? We have been doing most everything on the Wiki, but I would prefer to do things in the repo.

File: codalab/codalabtools/compute/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+psutil==4.0.0
+PyYAML==3.10
+
+
+# Added for python sdk
+git+https://github.com/FlavioAlexander/azure-sdk-for-python.git@legacy
```
Comment (ckcollab) [inline codalab/codalabtools/compute/requirements.txt]: I think once we merge in develop we can remove this, compute worker doesn't need Azure any more.

File: codalab/codalabtools/compute/worker.py [hunk 4]
```diff
@@ -219,6 +219,7 @@ def _send_update(task_id, status, secret, virtual_host='/', extra=None):
         # We need to send on the main virtual host, not whatever host we're currently
         # connected to.
         new_connection.virtual_host = virtual_host
+        print new_connection
```
Comment (ckcollab) [inline codalab/codalabtools/compute/worker.py]: Remove this

File: codalab/codalabtools/compute/worker.py [hunk 5]
```diff
@@ -28,9 +28,9 @@
 from subprocess import Popen, call
 from zipfile import ZipFile
 
-# Add codalabtools to the module search path
 from celery.app import app_or_default
 
+# Add codalabtools to the module search path
```
Comment (ckcollab) [inline codalab/codalabtools/compute/worker.py]: Thanks for fixing this!

File: docker-compose.yml [hunk 6]
```diff
@@ -1,9 +1,102 @@
...
+      # - AMQP_ADMIN_USERNAME=admin
+      # - AMQP_ADMIN_PASSWORD=rabbitmq
+      # - AMQP_ADMIN_HOST=rabbit
+      # - AMQP_ADMIN_PORT=15672
+    volumes:
+      - .:/app
+    depends_on:
+      - rabbit
+  
+  # Memcached Container
```
Comment (ckcollab) [inline docker-compose.yml]: should make this name consistent with the rest

File: docker-compose.yml [hunk 7]
```diff
@@ -1,9 +1,102 @@
...
+    ports:
+      - "8000:8000"
+    links:
+      - db
+      - rabbit
+      - memcached
+    depends_on:
+      - db
+
+  # Celery Workers
```
Comment (ckcollab) [inline docker-compose.yml]: Side note, any way to view old logs?

File: docker-compose.yml [hunk 8]
```diff
@@ -1,9 +1,102 @@
...
+    volumes:
+      - .:/app
+    links:
+      - db
+      - rabbit
+    depends_on:
+      - rabbit
+      - db
+    
+
```
Comment (ckcollab) [inline docker-compose.yml]: only 1 new line at end of file

File: run_compute.sh [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+mkdir -p /tmp/codalab
+chmod ugo+rwx /tmp/codalab
+
+mkdir -p /tmp/python-eggs
+export PYTHON_EGG_CACHE=/tmp/python-eggs
+chmod ugo+rwx /tmp/python-eggs
+
+# Start compute worker
+su -m myuser -c "celery -A codalab worker -l info -Q compute-worker -n compute-worker --concurrency=1 -Ofast -Ofair"
```
Comment (ckcollab) [inline run_compute.sh]: missing newline at end of file

File: run_web.sh [hunk 10]
```diff
@@ -0,0 +1,19 @@
+#!/bin/sh
+
+# wait for MYSQL server to start
+sleep 45
```
Comment (ckcollab) [inline run_web.sh]: can we make these smarter?

File: Dockerfile [hunk 11]
```diff
@@ -0,0 +1,8 @@
+FROM python:2.7
```
Comment (ckcollab) [inline Dockerfile]: Another side note: can we deploy this somewhere to test out everything, like a free AWS server or something?

File: docker-compose.yml (hunk 7 above)
Comment (jared-cain) [inline docker-compose.yml]: so long as you dont docker rm a container you should be able to view old logs with docker logs or by accessing the container directly

Comment (jared-cain) [inline docker-compose.yml]: so long as you dont "docker rm" a container you can view all of its logs by default with "docker logs"

File: run_web.sh (hunk 10 above)
Comment (jared-cain) [inline run_web.sh]: Yes. I am looking into a program called "wait-for-it".sh and that could be the key

File: Dockerfile (hunk 11 above)
Comment (jared-cain) [inline Dockerfile]: i believe so... i have done simple docker deploys on heroku so ill dive into this

File: run_compute.sh (hunk 9 above)
Comment (jared-cain) [inline run_compute.sh]: fixed in new push

File: docker-compose.yml [hunk 12]
```diff
@@ -1,9 +1,100 @@
...
+    image: totem/celery-flower-docker
+    hostname: flower
+    ports:
+      - "5555:5555"
+    environment:
+      - AMQP_USERNAME=guest
+      - AMQP_PASSWORD=guest
+      - AMQP_HOST=rabbit
+      - AMQP_PORT=5672
+      # - AMQP_ADMIN_USERNAME=admin
```
Comment (ckcollab) [inline docker-compose.yml]: pull this extra stuff?

File: docker-compose.yml (hunk 7 above)
Comment (ckcollab) [inline docker-compose.yml]: cool

File: run_flower.sh [hunk 13]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+su -m myuser -c "celery flower --broker=pyamqp://admin:rabbitmq@rabbit:5762//"
```
Comment (ckcollab) [inline run_flower.sh]: does this need to change to `guest:guest`? How will this work when we run on production, I wonder? Will it have to grab user/pass from env vars or something?

File: run_web.sh [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+  # wait for 5 seconds before check again
+  sleep 5
+done
+
+echo "WEB IS RUNNING"
+
+mkdir -p /tmp/codalab
+chmod ugo+rwx /tmp/codalab
+
+# mkdir -p /tmp/python-eggs
```
Comment (ckcollab) [inline run_web.sh]: pull these?

File: docker-compose.yml (hunk 12 above)
Comment (jared-cain) [inline docker-compose.yml]: yeah thats for the flower configuration to work

Comment (jared-cain) [inline docker-compose.yml]: the admin stuff CAN go away though?

File: run_flower.sh (hunk 13 above)
Comment (jared-cain) [inline run_flower.sh]: shoot good looks at this one... i will test out with just env vars real quick!

File: run_web.sh (hunk 14 above)
Comment (jared-cain) [inline run_web.sh]: roger that

File: run_web.sh [hunk 15]
```diff
@@ -0,0 +1,21 @@
+#!/bin/sh
+
+# wait for MYSQL server to start
+
+until netcat -z -v -w30 db 3306
```
Comment (ckcollab) [inline run_web.sh]: Can we make this work different, like, not require a port or something? Or grab the port from an ENV var?

File: README.md [hunk 16]
```diff
@@ -15,12 +15,29 @@ To see Codalab Competition's in action, visit [www.competitions.codalab.org](htt
...
 - [CodaLabDev Google Groups Forum](https://groups.google.com/forum/#!forum/codalabdev)
 
+
+## Compute worker
+
+To start the compute worker under project directory:
+
+`docker-compose up worker`
+
+If for some reason you need to rebuild the image run:
```
Comment (ckcollab) [inline README.md]: What would "some reasons" be here? Can you give some examples?

File: run_web.sh (hunk 15 above)
Comment (ckcollab) [inline run_web.sh]: Also can we change to `sleep 1`

**Comment by ckcollab** [comment]
Also, why is CircleCI hanging on these tests??

File: docker-compose.yml (hunk 7 above)
Comment (xbaro) [inline docker-compose.yml]: We use to map the log folder to an external volume, in this way we can access them from outside and they are persisted.

File: codalab/codalabtools/compute/Dockerfile [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+WORKDIR /app/codalab/codalabtools/compute
+RUN apt-get update && apt-get install -y \ 
+    build-essential psmisc
+RUN pip install -r requirements.txt
+RUN mkdir /tmp/codalab
+RUN chmod -R ugo+rwx /tmp/codalab
+ENTRYPOINT ["python"]
+CMD ["worker.py"]
+
+
```
Comment (ckcollab) [inline codalab/codalabtools/compute/Dockerfile]: extra newlines

File: codalab/codalabtools/compute/worker.py [hunk 18]
```diff
@@ -426,7 +426,7 @@ def run(task_id, task_args):
                 exit_code = None
                 timed_out = False
 
-                if 'Darwin' not in platform.platform():
+                if 'Darwin' not in platform.platform() and os.environ.get("DJANGO_ENVIRONMENT") is 'test':
```
Comment (ckcollab) [inline codalab/codalabtools/compute/worker.py]: did we want to revert this?

File: docker-compose.yml [hunk 19]
```diff
@@ -1,9 +1,96 @@
...
     ports:
-        - 3306:3306
+      - 3306:3306
+  
+  # RabbitMQ - rabbit
+  rabbit:
+    image: rabbitmq:3.5.7
+    hostname: rabbit
+    environment:
+      - RABBITMQ_DEFAULT_USER=guest
```
Comment (ckcollab) [inline docker-compose.yml]: are the env changes on another branch?

**Pull request merged**